docs/architecture/adr-032-typed-events.md

```diff
@@ -0,0 +1,313 @@
+    uint64 proposal_id    = 2;
+    TextProposal proposal = 3;
+}
+```
+
+__Step-3__: Refactor event emission to use the typed event created and emit using `sdk.EmitTypedEvent`:
+
+```go
+// x/gov/handler.go
+func handleMsgSubmitProposal(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgSubmitProposalI) (*sdk.Result, error) {
+    ...
+    types.Context.EventManager().EmitTypedEvent(
+        &EventSubmitProposal{
+            FromAddress: fromAddress,
+            ProposalId: id,
+            Proposal: proposal,
+        },
+    )
+    ...
+}
+```
+
+#### How to subscribe to these typed events in `Client`
+
+> NOTE: Full code example below
+
+Users will be able to subscribe using `client.Context.Client.Subscribe` and consume events which are emitted using `EventHandler`s.
+
+Akash Network has built a simple [`pubsub`](https://github.com/ovrclk/akash/blob/master/pubsub/bus.go). This can be used to subscribe to `abci.Events` and [publish](https://github.com/ovrclk/akash/blob/master/events/publish.go#L21) them as typed events.
+
+Please see the below code sample for more detail on this flow looks for clients.
+
+## Consequences
+
+### Positive
+
+* Improves consistency of implementation for the events currently in the sdk
+* Provides a much more ergonomic way to handle events and facilitates writing event driven applications
+* This implementation will support a middleware ecosystem of `EventHandler`s
+
+### Negative
+
+
+## Detailed code example of publishing events
+
+This ADR also proposes adding affordances to emit and consume these events. This way developers will only need to write
+`EventHandler`s which define the actions they desire to take. 
+
+```go
+// EventEmitter is a type that describes event emitter functions
+// This should be defined in `types/events.go`
+type EventEmitter func(context.Context, client.Context, ...EventHandler) error
+
+// EventHandler is a type of function that handles events coming out of the event bus
+// This should be defined in `types/events.go`
+type EventHandler func(proto.Message) error
+
+// Sample use of the functions below
+func main() {
+    ctx, cancel := context.WithCancel(context.Background())
+
+    if err := TxEmitter(ctx, client.Context{}.WithNodeURI("tcp://localhost:26657"), SubmitProposalEventHandler); err != nil {
+        cancel()
+        panic(err)
+    }
+
+    return
+}
+
+// SubmitProposalEventHandler is an example of an event handler that prints proposal details
+// when any EventSubmitProposal is emitted. 
+func SubmitProposalEventHandler(ev proto.Message) (err error) {
+    switch event := ev.(type) {
+    // Handle governance proposal events creation events
+    case govtypes.EventSubmitProposal:
+        // Users define business logic here e.g.
+        fmt.Println(ev.FromAddress, ev.ProposalId, ev.Proposal)
+        return nil
+    default:
+        return nil
+    }
+}
+
+// TxEmitter is an example of an event emitter that emits just transaction events. This can and 
```

Comment:
> This can and should be implemented somewhere in the SDK.

This is a huge architectural change unrelated to the typed-event part. Currently all indexing is in tendermint and the SDK doesn't deal with those other event. I would pull such proposed changes into a different ADR and get lots of input from the tendermint team on splitting responsibilities. 

*Remember: Tendermint is not just for the Cosmos SDK... and not just for Go applications*

Reply:
@ethanfrey this is an SDK only change that would only impact how to deal with these events in Go, and no other language. Its a pretty bounded change that can be easily implemented in a 3rd party lib, but would be nice to haves in the SDK.